packages/firestore/src/local/local_store.ts

```diff
@@ -642,6 +615,191 @@ export class LocalStore {
       });
   }
 
+  /**
+   * Applies the documents from a bundle to the "ground-state" (remote)
+   * documents.
+   *
+   * LocalDocuments are re-calculated if there are remaining mutations in the
+   * queue.
+   */
+  applyBundledDocuments(
+    documents: BundledDocuments
+  ): Promise<MaybeDocumentMap> {
+    let updatedKeys = documentKeySet();
+    let documentMap = maybeDocumentMap();
+    let versionMap = documentVersionMap();
+    for (const [metadata, doc] of documents) {
+      const documentKey = this.bundleConverter.toDocumentKey(metadata.name!);
+      updatedKeys = updatedKeys.add(documentKey);
+      documentMap = documentMap.insert(
+        documentKey,
+        this.bundleConverter.toMaybeDocument(metadata, doc)
+      );
+      versionMap = versionMap.insert(
+        documentKey,
+        this.bundleConverter.toSnapshotVersion(metadata.readTime!)
+      );
+    }
+
+    const documentBuffer = this.remoteDocuments.newChangeBuffer({
+      trackRemovals: true // Make sure document removals show up in `getNewDocumentChanges()`
+    });
+    return this.persistence.runTransaction(
+      'Apply bundle documents',
+      'readwrite-primary',
+      txn => {
+        return this.applyDocuments(
+          documentBuffer,
+          txn,
+          updatedKeys,
+          documentMap,
+          versionMap
+        )
+          .next(changedDocs => {
+            documentBuffer.apply(txn);
+            return changedDocs;
+          })
+          .next(changedDocs => {
+            return this.localDocuments.getLocalViewOfDocuments(
+              txn,
+              changedDocs
+            );
+          });
+      }
+    );
+  }
+
+  /**
+   * Applies documents to remote document cache, returns the document changes
+   * resulting from applying those documents.
+   *
+   * @param updatedKeys Keys of the documents to be applied.
+   * @param documents Documents to be applied.
+   * @param remoteVersion The read time of the documents to be applied, it is
+   * a `SnapshotVersion` if all documents have the same read time, or a `DocumentVersionMap` if
+   * they have different read times.
+   */
+  private applyDocuments(
+    documentBuffer: RemoteDocumentChangeBuffer,
+    txn: PersistenceTransaction,
+    updatedKeys: DocumentKeySet,
+    documents: MaybeDocumentMap,
+    remoteVersion: SnapshotVersion | DocumentVersionMap
```

Comment:
Please split this up into two arguments: `gloablSnapshotVersion` and `doucmentVersions`. It will make the code below a lot easier to follow. With my advice above, you now even get to keep the same number of arguments for this method :)